diff -ur source-orig/.build/compile.step source/.build/compile.step
--- source-orig/.build/compile.step	1969-12-31 19:00:01.000000000 -0500
+++ source/.build/compile.step	2019-07-02 15:10:35.892764178 -0400
@@ -69,7 +69,7 @@
           failonerror="false"
           if="${platform::is-windows()}" />
     <exec
-      program="mono"
+      program="mono"
       workingdir="${dirs.build}"
       failonerror="false"
       if="${platform::is-unix()}"
